xbmc/utils/UnicodeConverter.cpp

```diff
@@ -0,0 +1,992 @@
+      { // is constexpr
+        if (TRACE_OUTPUT) // TODO: remove before merge. Helps with debugging on
+        { // Jenkins build machines.
+          std::cout << "iconv E2BIG input: " << StringUtils::ToHex(input) << std::endl;
+          std::cout << "output: " << StringUtils::ToHex(output) << std::endl;
+          std::cout << "inCursor: " << inCursor << std::endl;
+        }
+        iconvRC = 0; // Not a serious problem. Just empty buffer and continue
+      }
+      else if (errno == EILSEQ) // An invalid multibyte sequence has been
+      { // encountered in the input. Note that iconv may not
+        // precisely define which is the offending byte, but
+        // it will be close. Also, we should acknowledge
+        // that, say, a u32string output type, may cause
+        // multiple bad characters identified, but this is
+        // likely due to an error occurring in, say, a
+        // u32string, where four bytes is the minimum
+        // code unit size. If an error occurs, it is
+        // impossible to identify which byte is responsible.
+        if (TRACE_OUTPUT)
+        {
+          std::cout << "iconv EILSEO input: " << StringUtils::ToHex(input) << std::endl;
+          std::cout << "output: " << StringUtils::ToHex(output) << std::endl;
+          std::cout << "inCursor: " << inCursor << " previousFailure: " << previousFailure
+                    << std::endl;
+        }
+        if (!failOnInvalidChar)
+        {
+          iconvRC = 0; // Don't fail on invalid char
+          if (substituteBadCodepoints and
+              !previousFailure) // Consider consecutive failures as one. Could skip to next start byte
+          {
+            // substitution occurs AFTER appending any good converted chars
+
+            if (TRACE_OUTPUT)
+              std::cout << "Append substitution codepoint" << std::endl;
+            appendSubChar = true;
+            previousFailure = true;
+          }
+
+          // Skip offending codeunit
+
+          inCursor += inBytesPerCU;
+          inBytesRemaining -= inBytesPerCU;
+        }
+        else // failed & failOnInvalidChar
+        {
+          inBytesRemaining = 0; // Bail out
+          outBufAvail = outBufByteSize; // Nothing to copy
+        }
+      }
+      else if (errno == EINVAL) // Invalid sequence at the end of input buffer. Either needs
+      { // more input, or the end is malformed. In our case, can not be the former.
+        // Insert substitute character, if chosen
+
+        if (TRACE_OUTPUT)
+        {
+          std::cout << "iconv EINVAL input: " << StringUtils::ToHex(input) << std::endl;
+          std::cout << "output: " << StringUtils::ToHex(output) << std::endl;
+          std::cout << "inCursor: " << inCursor << " previousFailure: " << previousFailure
+                    << std::endl;
+        }
+        if (!failOnInvalidChar)
+        {
+          iconvRC = 0; // Don't fail on invalid char
+
+          if (substituteBadCodepoints and
```

Comment:
```suggestion
          if (substituteBadCodepoints &&
```
do not use alternative operators